spec/helpers/faker/puppet_module_remote_repo.rb

```diff
@@ -0,0 +1,140 @@
+require 'open3'
+
+require_relative '../faker'
+
+module ModuleSync
+  # Fake a remote git repository that holds a puppet module
+  #
+  # This module allows to fake a remote repositiory using:
+  #  - a bare repo
+  #  - a temporary cloned repo to operate on the remote before exposing to modulesync
+  #
+  # Note: This module needs to have working_directory sets before using it
+  module Faker
+    class PuppetModuleRemoteRepo
+      class CommandExecutionError < StandardError; end
+
+      attr_reader :name
+      attr_reader :namespace
+
+      def initialize(name, namespace)
+        @name = name
+        @namespace = namespace
+      end
+
+      def populate
+        FileUtils.chdir(Faker.working_directory) do
+          run %W{git init --bare #{bare_repo_dir}}
+          run %W{git clone #{bare_repo_dir} #{tmp_repo_dir}}
+
+          module_short_name = name.split('-').last
+
+          FileUtils.chdir(tmp_repo_dir) do
+            metadata = {
+              name: "modulesync-#{module_short_name}",
+              version: '0.4.2',
+              author: 'ModuleSync team',
+            }
+
+            File.write 'metadata.json', metadata.to_json
+            run %w{git add metadata.json}
+            run %W{git commit --message} << 'Initial commit'
+            run %w{git push}
+          end
+        end
+      end
+
+      def read_only=(value)
+        mode = value ? '0444' : '0644'
+        FileUtils.chdir(bare_repo_dir) do
+          run %W{git config core.sharedRepository #{mode}}
+        end
+      end
+
+      def default_branch
+        FileUtils.chdir(bare_repo_dir) do
+          stdout = run %w{git symbolic-ref --short HEAD}
+          return stdout.chomp
+        end
+      end
+
+      def default_branch=(value)
+        FileUtils.chdir(bare_repo_dir) do
+          run %W{git branch -M #{default_branch} #{value}}
+          run %W{git symbolic-ref HEAD refs/heads/#{value}}
+        end
+      end
+
+      def read_file(filename, branch = nil)
+        branch ||= default_branch
+        FileUtils.chdir(bare_repo_dir) do
+          return run %W{git show #{branch}:#{filename}}
```

Comment:
Is this a good idea with variables? Isn't a real array safer?
```suggestion
          return run ['git', 'show', "#{branch}:#{filename}"]
```
Otherwise I'm worried that it ends up splitting spaces in filenames into separate arguments.

Also looks like this is a repeated pattern

Reply:
```irb
irb(main):001:0> with_space = 'Hello world'
irb(main):002:0> without_space = 'NoSpace'
irb(main):003:0> full = %W[#{without_space}:#{with_space}]
irb(main):004:0> full
=> ["NoSpace:Hello world"]
```

Using a recent rubocop against this file, it only asked me to use `%w[]` instead of `%w{}` e.g.:
```
spec/helpers/faker/puppet_module_remote_repo.rb:50:15: C: [Corrected] Style/PercentLiteralDelimiters: %W-literals should be delimited by [ and ].
```

Reply:
I guess I learned something new today. I still find Ruby's many ways to do the same thing confusing.